src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,123 @@
+        private const string ImmutableStackGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableStack`1";
+
+        private const string ImmutableQueueGenericTypeName = "System.Collections.Immutable.ImmutableQueue`1";
+        private const string ImmutableQueueGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableQueue`1";
+
+        private const string ImmutableSortedSetGenericTypeName = "System.Collections.Immutable.ImmutableSortedSet`1";
+
+        private const string ImmutableHashSetGenericTypeName = "System.Collections.Immutable.ImmutableHashSet`1";
+        private const string ImmutableSetGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableSet`1";
+
+        // This method is invoked with reflection in CreateFromList.
+        private static IEnumerable CreateFromListInternal<T>(MethodInfo createImmutableMethod, IList sourceList)
+        {
+            List<T> list = new List<T>();
+
+            foreach (object item in sourceList)
+            {
+                if (item is T itemT)
+                {
+                    list.Add(itemT);
+                }
+            }
+
+            return (IEnumerable)createImmutableMethod.Invoke(null, new object[] { list } );
+        }
+
+        public override IEnumerable CreateFromList(Type enumerableType, Type elementType, IList sourceList)
+        {
+            Debug.Assert(enumerableType.IsGenericType);
+
+            Type underlyingType = enumerableType.GetGenericTypeDefinition();
+            Type constructingType;
+            string name = underlyingType.FullName;
+
+            switch (name)
+            {
+                case ImmutableListGenericTypeName:
+                case ImmutableListGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableList);
+                    break;
+                case ImmutableArrayGenericTypeName:
+                    constructingType = typeof(ImmutableArray);
+                    break;
+                case ImmutableStackGenericTypeName:
+                case ImmutableStackGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableStack);
+                    break;
+                case ImmutableQueueGenericTypeName:
+                case ImmutableQueueGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableQueue);
+                    break;
+                case ImmutableSortedSetGenericTypeName:
+                    constructingType = typeof(ImmutableSortedSet);
+                    break;
+                case ImmutableHashSetGenericTypeName:
+                case ImmutableSetGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableHashSet);
+                    break;
+                default:
+                    return sourceList;
+            }
+
+            MethodInfo[] constructingTypeMethods = constructingType.GetMethods();
+            MethodInfo constructingMethod = null;
```

Comment:
Use following LINQ instead of lines 88-97?

`MethodInfo constructingMethod = constructingTypeMethods.FirstOrDefault(m => m.Name == "CreateRange" && m.GetParameters().Length == 1);`

Reply:
Linq is slower and allocates a bit more, so I wouldn't make this change unless perf isn't critical here:
![image](https://user-images.githubusercontent.com/6527137/57109973-ca614f00-6ceb-11e9-9a1a-b6effefe922b.png)

```C#
private MethodInfo[] _constructingTypeMethods;

[GlobalSetup]
public void Setup()
{
    Type constructingType = typeof(ImmutableList);
    _constructingTypeMethods = constructingType.GetMethods();
}

[Benchmark(Baseline = true)]
public MethodInfo Linq()
{
    MethodInfo constructingMethod = _constructingTypeMethods.FirstOrDefault(m => m.Name == "CreateRange" && m.GetParameters().Length == 1);
    return constructingMethod;
}

[Benchmark]
public MethodInfo ForEach()
{
    MethodInfo constructingMethod = null;

    foreach (MethodInfo method in _constructingTypeMethods)
    {
        if (method.Name == "CreateRange" && method.GetParameters().Length == 1)
        {
            constructingMethod = method;
            break;
        }
    }

    return constructingMethod;
}
```

Reply:
I'll leave as is.